Dynamically look up the global defaultColor

etimberg opened the pull request:
This allows the colour to be programmatically changed. Additionally, this change ensures that `helpers.valueOrDefault` is used inside each element.

Resolves #4299 

This PR: https://codepen.io/anon/pen/vvgPPp
This PR with global element setting: https://codepen.io/anon/pen/VqpMvg
v2.7.3: https://codepen.io/anon/pen/pqRYBM

nagix commented on src/elements/element.point.js:
```diff
@@ -75,16 +71,17 @@ module.exports = Element.extend({
-			ctx.lineWidth = helpers.valueOrDefault(vm.borderWidth, globalDefaults.elements.point.borderWidth);
-			ctx.fillStyle = vm.backgroundColor || defaultColor;
+			ctx.strokeStyle = valueOrDefault(vm.borderColor, defaultColor);
+			ctx.lineWidth = valueOrDefault(vm.borderWidth, globalDefaults.elements.point.borderWidth);
```
Same as above

etimberg commented on src/elements/element.line.js:
```diff
@@ -43,17 +40,17 @@ module.exports = Element.extend({
+		ctx.lineDashOffset = valueOrDefault(vm.borderDashOffset, globalOptionLineElements.borderDashOffset);
+		ctx.lineJoin = valueOrDefault(vm.borderJoinStyle, globalOptionLineElements.borderJoinStyle);
+		ctx.lineWidth = valueOrDefault(vm.borderWidth, globalOptionLineElements.borderWidth);
+		ctx.strokeStyle = valueOrDefault(vm.borderColor, globalDefaults.defaultColor);
```
Here I think we might need to have some helper that falls back multiple times: `valueOrDefault(vm.borderColor, globalOptionLineElements.borderColor, globalDefaults.defaultColor)`.

nagix replied:
`helpers.options.resolve` can be used here, but are multiple `? :` operators better as this is drawing method? 
```js
resolve([vm.borderColor, globalOptionLineElements.borderColor, globalDefaults.defaultColor]);
```

kurkle replied:
Ideally those should be resolved before drawing, but for now I would leave it as is (for performance)

benmccann replied:
It seems fine to me to merge this since it doesn't change the behavior at all. If we want to investigate moving some logic to the controller, I think that would best be handled separately

etimberg commented on src/elements/element.line.js:
```diff
@@ -43,17 +40,17 @@ module.exports = Element.extend({
 
 		// Stroke Line Options
-		ctx.lineCap = vm.borderCapStyle || globalOptionLineElements.borderCapStyle;
+		ctx.lineCap = valueOrDefault(vm.borderCapStyle, globalOptionLineElements.borderCapStyle);
```
I think the fallback is mostly for the case where an invalid value is set. I'd be a bit wary about remove the fallback here since it has existed for a while

simonbrunel replied:
> I think the fallback is mostly for the case where an invalid value is set

I think elements should not have to deal with the options fallback, only controllers should implement that logic. If for any reason the element model contains `undefined` values, it should be considered a bug or it's expected and the element knows how to handle this undefined value (e.g. `lineDashOffset === undefined` could mean no dash). 

We should maybe try to remove these defaults fallback from elements since it silents bugs at the controller level?

Or is it really too risky?

etimberg replied:
This might be a good reason to close this PR without merging it. Instead we'd do it at the controller level. However that would mean that changing the `defaultColor` would require calling `update()` afterwards. I think that's acceptable to me

nagix replied:
I think we still need to remove the line below (and similar lines) otherwise the changes in global defualt options won’t be applied when `update` is called.
https://github.com/chartjs/Chart.js/blob/e07c5d0c6c39e621fbe79de39d6e2de66da1cd9a/src/elements/element.line.js#L15

kurkle replied:
I agree with @simonbrunel, we should remove this fallback from elements.
I also agree with @etimberg in that its ok to require `update` after change in defaults.

Considering that all internal controllers have new fallback logic after 2.7.3, fallback removal in elements should make no difference.

So I'd either remove the fallback or revert to using `||` if removal seems too risky.

etimberg replied:
I removed the fallbacks and did this at entire change at the resolve level

kurkle commented on src/elements/element.line.js:
```diff
@@ -43,17 +40,17 @@ module.exports = Element.extend({
 
 		// IE 9 and 10 do not support line dash
 		if (ctx.setLineDash) {
 			ctx.setLineDash(vm.borderDash || globalOptionLineElements.borderDash);
```
shouldn't this line be converted to `valueOrDefault` too?

etimberg replied:
I can make that change

benmccann commented:
@etimberg this change will have to be rebased

benmccann commented:
@etimberg this PR will need to be rebased

benmccann commented:
@etimberg just a reminder that this PR needs a rebase

benmccann commented:
@etimberg just a gentle reminder that this PR needs to be rebased

benmccann commented on src/elements/element.arc.js:
```diff
@@ -89,14 +88,17 @@ module.exports = Element.extend({
 		var pixelMargin = (vm.borderAlign === 'inner') ? 0.33 : 0;
 		var angleMargin;
 
+		var defaultColor = defaults.global.defaultColor;
```
We've started placing aliased methods at the top of the file

etimberg replied:
The challenge with that is if we capture it, someone creates a chart, and then changes the default color we'll never get the new value

benmccann commented:
@etimberg this PR will need to be rebased again

kurkle commented on src/elements/element.line.js:
```diff
@@ -45,17 +40,17 @@ module.exports = Element.extend({
 		// IE 9 and 10 do not support line dash
 		if (ctx.setLineDash) {
-			ctx.setLineDash(vm.borderDash || globalOptionLineElements.borderDash);
+			ctx.setLineDash(valueOrDefault(vm.borderDash, globalOptionLineElements.borderDash));
```
Reconsidering this; many additional function calls for each data point. I think we should keep using `||` in `draw`, because these are performance critical functions.

etimberg replied:
Removed the fallbacks entirely

etimberg commented:
I took a new approach on this. I removed the element level fallbacks entirely and updated the resolve step of the controllers to fallback to `global.defaults.defaultColor` during that process.

I removed a few tests that explicitly tested that behaviour. I had to do a lot of updates to the line element tests that use the fake canvas. I think all of that code is tested via the image based tests but I didn't want to expand the scope of this PR to handle that. I'll look into that in a separate PR.

benmccann commented:
@etimberg this PR will need to be rebased

benmccann commented:
@etimberg do you plan to come back to this PR? I'm wondering if we should close it for the time being until you get a chance to take another look so that we keep the review queue clean

etimberg commented:
I tried to rebase this, but so much has changed that it was getting pretty complicated. I'm going to close it for now and we can always come back to it
